heapster-saw/doc/tutorial/tutorial.md

```diff
@@ -0,0 +1,1265 @@
+implication are identical, since the function doesn't change the shape
+of memory. The return value is `void`, so we can omit it or add a
+trivial `ret:true`.
+
+The permission for pointers `ptr` takes three arguments. First, it
+describes the read-write modality. In this case the
+pointer is writable `W`, since it will be modified. The second
+argument describes the pointer offset, here `0`. Finaly, the third
+argument describes the content of the pointer, in this case a 64-bit
+integer `int64<>`.
+
+Then we can type-check the function with
+
+```
+heapster_typecheck_fun env "incr_ptr" "(). arg0:ptr((W,0) |-> int64<>) -o arg0:ptr((W,0) |-> int64<>)"
+```
+
+Finally we can generate the functional specification in Coq with 
+
+```
+heapster_export_coq env "tutorial_c_gen.v";
+```
+
+The old file should be overwritten and now contains the functional
+specification of `add`, `add_misstyped` and `incr_ptr`. As you can see
+the definition of `incr_ptr__tuple_fun` has no references to `errorM`,
+so we know it was correctly type checked.
+
+You will have to generate the `.vo` again to write proofs about
+`incr_ptr`. After you do so, we can easily prove that `incr_ptr`
+produces no errors.
+
+```
+Lemma no_errors_incr_ptr
+  : refinesFun incr_ptr (fun _ => noErrorsSpec).
+Proof.
+  unfold incr_ptr, incr_ptr__tuple_fun.
+  prove_refinement.
+Qed.
+```
+
+### Structs
+
+The next function we will type-check deals with structs. In our
+example, we defined a function that can compute the norm of a 3D
+vector
+
+``` C
+// Struct that represents the three coordinates fora 3D vector
+typedef struct { uint64_t x; uint64_t y; uint64_t z; } vector3d;
+
+// function that computes the norm of a 3D vector
+// || (x,y,z) || = x^2+y^2+z^2
+uint64_t norm_vector (vector3d *v) { return (v->x * v->x + v->y * v->y + v->z * v->z); }
+```
+
+Again, we assume that you still have the definitions from the previous
+sections so we can start defining the type for the function.
+
+Let's start by defining a predicate for `vector3d` like so
+
+```
+heapster_define_perm env "vec3d" "rw:rwmodality" "llvmptr 64" "ptr((rw,0) |-> int64<>) * ptr((rw,8) |-> int64<>) * ptr((rw,16) |-> int64<>)"
+```
+
+First, notice that we added an arguments `rw` of type
+`rwmodality`. This is such that we can control if the vector is
+readable or writable. Second, notice that the predicate still applies
+to 64-bit pointers. Finally, the type describes three integers, all
+with the read-write modality given by the argument `rw`, at offsets
+`0`, `8` and `16` and each with an `int64`.
+
+Then we can define the type of `norm_vector` as
+
+```
+(). arg0:vec3d<R> -o arg0:vec3d<R>, ret:int64<>
+```
+
+which says that the function takes a readable 3D vector, and returns
+an integer. Notice that the `arg0` on the right hand side could also
+be written as `arg0:true`. However, we still want to express that the
+function does change that memory so we make it explicit.
```

Comment:
Wait, why does the function change the memory? I wouldn't have expected that since it's using the `R` modality.

Reply:
Actually, yes, `norm_vector` should just have `arg0:true` as its output permission. The convention is that any copyable permissions in the input, such as `eq` or read-only pointer permissions, do not need to get returned in the output.